src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+            return (
+                dec_features,
+                dec_hidden,
+                encoder_out.encoder_out,
+                encoder_out.encoder_states,
+            )
+        elif self.output_attentions:
+            return (dec_features, dec_attn, encoder_out.encoder_out, encoder_out.encoder_attn)
+        else:
+            return (dec_features, encoder_out.encoder_out)
+
+    @staticmethod
+    def shift_tokens_left(input_ids, pad_token_id):
+        """Shift input ids one token to the left"""
+        prev_output_tokens = input_ids.clone()
+        prev_output_tokens[:, 0] = input_ids.gather(
+            1, (input_ids.ne(pad_token_id).sum(dim=1) - 1).unsqueeze(-1),
+        ).squeeze()
+        prev_output_tokens[:, 1:] = input_ids[:, :-1]
+        return prev_output_tokens
+
+    def max_positions(self):
+        """Maximum length supported by the model."""
+        return (self.encoder.max_positions(), self.decoder.max_positions())
+
+
+class BartForMaskedLM(PretrainedBartModel):
+    base_model_prefix = "model"
+
+    def __init__(self, config: BartConfig):
+        super().__init__(config)
+        self.model = BartModel(config)
+
+    def forward(self, *args, **kwargs):  # TODO(SS): better signature when base API decided
+        kwargs["return_for_head"] = True
+        lm_labels = kwargs.pop("lm_labels", None)
+        decoder_outputs, encoder_outputs = self.model(*args, **kwargs)
+        lm_logits = self.output_layer(decoder_outputs[0])
+        decoder_outputs = (lm_logits,) + decoder_outputs[1:]  # Add hidden states and attention if they are here
+
+        if lm_labels is not None:
+            loss_fct = nn.CrossEntropyLoss()
+            masked_lm_loss = loss_fct(lm_logits.view(-1, self.config.vocab_size), lm_labels.view(-1))
+            decoder_outputs = (masked_lm_loss,) + decoder_outputs
+        return decoder_outputs + encoder_outputs
+
+    def output_layer(self, x):
+        return F.linear(x, self.model.shared.weight)
+
+
+class BartForSequenceClassification(PretrainedBartModel):
+    eos_token = 2
+
+    def __init__(self, config: BartConfig, **kwargs):
+        super().__init__(config, **kwargs)
+        self.model = BartModel(config)
+        self.classification_head = BARTClassificationHead(
+            config.d_model, config.d_model, config.num_labels, config.classif_dropout,
+        )
+        self.loss_fn = nn.CrossEntropyLoss()
+
+    def forward(self, input_ids, *args, **kwargs):
+
+        if input_ids.ndim == 1:
+            input_ids = input_ids.unsqueeze(0)
+        kwargs["return_for_head"] = True
+        labels = kwargs.pop("labels", None)
+        decoder_outputs, encoder_outputs = self.model(input_ids, *args, **kwargs)
+        x = decoder_outputs[0]  # last hidden state
+
+        eos_mask = input_ids.eq(self.eos_token)
+        if len(torch.unique(eos_mask.sum(1))) > 1:
+            raise ValueError("All examples must have the same number of <eos> tokens.")
```

Comment:
Up to now we have avoided mixing model and tokenization logic to let the user more free in particular. we should discuss whether we want to change this philosophy here. cc @julien-c @LysandreJik

Reply:
I'm still a firm believer that model logic and tokenization logic should be separate. We used to have a similar warning/error in RoBERTa but removed it because it 

1) made it impossible to trace the method, therefore not usable on TPU 
2) dissociated it from the traditional tokenizer logic =/= model logic that is used in all other models

Reply:
I think the best way is to use `SequenceSummary(cls_index)`. Still might need the eos token to make the `cls_index` tensor, if I'm understanding it correctly.

Reply:
Yes, sounds reasonable.